lib/alp_workloads/kvm_workload_utils.pm

```diff
@@ -0,0 +1,201 @@
+    save_screenshot;
+    assert_screen('in-libvirtd-container-bash');
+    assert_script_run("echo Hello from libvirtd container");
+    save_screenshot;
+
+    assert_script_run("ip a");
+    save_screenshot;
+    record_info('Please check if it is from within libvirtd container');
+
+}
+
+sub exit_kvm_container {
+    assert_script_run("clear");
+    assert_screen('in-libvirtd-container-bash');
+    enter_cmd("exit");
+    wait_still_screen 3;
+    save_screenshot;
+    assert_screen('back-to-host-shell');
+
+}
+
+sub install_tools_within_kvm_container {
+    my $_tools_to_install = shift;
+
+    $_tools_to_install = "curl openssh hostname gawk supportutils" if (!$_tools_to_install);
+
+    assert_script_run("clear");
+    assert_screen('in-libvirtd-container-bash');
+    zypper_call("in $_tools_to_install", 120);
+    save_screenshot;
+    record_info("Successfully installed test needed tools within VT container.");
+
+}
+
+sub setup_services_within_kvm_container {
+    assert_script_run("clear");
+    assert_screen('in-libvirtd-container-bash');
+
+    # Stop all pre-start virtual network
+    clean_all_virt_networks;
+    # Setup nat virtual network as needed
+    my $_vnet_loc = "/etc/libvirt/qemu/networks/alp-nat-vnet.xml";
+    assert_script_run("curl " . data_url("virt_autotest/alp-nat-vnet.xml") . " -o $_vnet_loc");
+    assert_script_run("virsh net-define $_vnet_loc");
+    assert_script_run("virsh net-autostart test-virt-net");
+    assert_script_run('virsh net-start test-virt-net');
+    save_screenshot;
+    record_info("Successfully setup nat virtual network test-virt-net for automation test.", script_output("ip a"));
+}
+
+sub setup_kvm_container_from_scratch {
+    pull_kvm_container;
+    config_host_and_kvm_container;
+    start_kvm_container;
+    enter_kvm_container_sh;
+    install_tools_within_kvm_container;
+    setup_services_within_kvm_container;
+}
+
+sub clean_and_resetup_kvm_container {
+    cleanup_host_and_kvm_container;
+    setup_kvm_container_from_scratch;
+}
+
+sub collect_kvm_container_setup_logs {
+    reset_consoles;
+    use_ssh_serial_console;
+
+    my $_kvm_container_log_file = "/tmp/kvm_container.log";
+    my @cmds = (
+        "$ct_engine ps --all",
+        "$ct_engine images -a --digests",
+        "$ct_engine inspect libvirtd",
+        "$ct_engine logs libvirtd"
+    );
+
+    script_run("rm $_kvm_container_log_file");
+    foreach my $cmd (@cmds) {
+        script_run("echo \"Executing $cmd:\" >>$_kvm_container_log_file");
+        script_run("$cmd >> $_kvm_container_log_file 2>&1");
+        script_run(qq@echo "\n\n" >> @ . $_kvm_container_log_file);
+    }
+    save_screenshot;
+    upload_logs("$_kvm_container_log_file");
+
+    upload_logs("/etc/kvm-container.conf");
+    upload_logs("/usr/local/bin/kvm-container-manage.sh");
+    upload_logs("/etc/libvirt/qemu/networks/alp-nat-vnet.xml");
+    save_screenshot;
+}
```

Comment:
Regarding "libvirtd container" log, I think you can create a systemd unit for "kvm container" by "podman generate systed", so "libvirtd container" can be controlled as well on host by systemctl. There will be service logs for this "libvirtd container" on host, can you can also make "kvm container" start up on system boot/reboot/power cycle by "systemctl enable". Very convenient and handy.

Reply:
It is already provided within kvm container repo as `https://build.opensuse.org/package/view_file/SUSE:ALP:Workloads/kvm-container/libvirtd.service?expand=1`. Will likely to be recommended to users by default(dev's plan). But now it is  a little buggy. So will add that later in our automation test code.